### PR TITLE
Revise and simplify the syntax model for out var

### PR DESCRIPTION
**Pull request by gafter**
The simplification is the introduction of a "declaration expression" in the context of an out argument.

### CODE REVIEW

**Comment by crummel** [comment]
Build triggered. sha1 is merged.

**Comment by gafter** [comment]
@dotnet/roslyn-compiler Please review.

**Comment by gafter** [comment]
FYI - this is likely to be revised slightly again. I am working on a proposed revision of the syntax for deconstruction in the context of #12664. Part of that will be replacing `VariableDeclaration` in `DeclarationExpression` with a new syntax form (being introduced as part of the deconstruction revision) that can represent a variable declaration `Type id`, or a deconstruction `var (x, y)` or `(int x, int y)`, or a wildcard `*`. When that is done the out var feature will be syntactically capable of representing all three.

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 1]
```diff
@@ -2101,19 +2099,21 @@ private bool RefMustBeObeyed(bool isDelegateCreation, ArgumentSyntax argumentSyn
 
         private BoundExpression BindArgumentValue(DiagnosticBag diagnostics, ArgumentSyntax argumentSyntax, bool allowArglist, RefKind refKind)
         {
-            if (argumentSyntax.Expression != null)
+            var declarationExpression = argumentSyntax.Expression as DeclarationExpressionSyntax;
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: We were just discussion type checks. In this case, it seems that a kind check is possible with `SyntaxKind.DeclarationExpression`. Would that be preferable?

File: src/Compilers/CSharp/Portable/PublicAPI.Unshipped.txt [hunk 2]
```diff
@@ -185,15 +190,16 @@ override Microsoft.CodeAnalysis.CSharp.Syntax.VariableDeconstructionDeclaratorSy
 override Microsoft.CodeAnalysis.CSharp.Syntax.WhenClauseSyntax.Accept(Microsoft.CodeAnalysis.CSharp.CSharpSyntaxVisitor visitor) -> void
 override Microsoft.CodeAnalysis.CSharp.Syntax.WhenClauseSyntax.Accept<TResult>(Microsoft.CodeAnalysis.CSharp.CSharpSyntaxVisitor<TResult> visitor) -> TResult
 static Microsoft.CodeAnalysis.CSharp.CSharpExtensions.GetDeclaredSymbol(this Microsoft.CodeAnalysis.SemanticModel semanticModel, Microsoft.CodeAnalysis.CSharp.Syntax.DeclarationPatternSyntax declarationSyntax, System.Threading.CancellationToken cancellationToken = default(System.Threading.CancellationToken)) -> Microsoft.CodeAnalysis.ISymbol
+static Microsoft.CodeAnalysis.CSharp.SyntaxExtensions.Identifier(this Microsoft.CodeAnalysis.CSharp.Syntax.DeclarationExpressionSyntax self) -> Microsoft.CodeAnalysis.SyntaxToken
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/PublicAPI.Unshipped.txt]: If I recall correctly, I thought we had left the discussion thinking we'd keep those two helpers internal for now.

File: src/Workspaces/CSharp/Portable/Extensions/SemanticModelExtensions.cs [hunk 3]
```diff
@@ -237,6 +232,10 @@ private static bool CanBindToken(SyntaxToken token)
                 {
                     current = ((CastExpressionSyntax)current).Expression;
                 }
+                else if (current is DeclarationExpressionSyntax)
```
Comment (jcouv) [inline src/Workspaces/CSharp/Portable/Extensions/SemanticModelExtensions.cs]: Could check on kind?

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs (hunk 1 above)
Comment (gafter) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: Yes, it would, especially since it is the uncommon case. Will revise.

File: src/Compilers/CSharp/Portable/PublicAPI.Unshipped.txt (hunk 2 above)
Comment (gafter) [inline src/Compilers/CSharp/Portable/PublicAPI.Unshipped.txt]: I thought I had changed them. Will change.

File: src/Workspaces/CSharp/Portable/Extensions/SemanticModelExtensions.cs (hunk 3 above)
Comment (gafter) [inline src/Workspaces/CSharp/Portable/Extensions/SemanticModelExtensions.cs]: Possible, but it would be inconsistent with the surrounding code. Probably this whole method could be refactored as a switch on Kind, but that is out of scope for this PR.

**Comment by gafter** [comment]
@jcouv I've responded to all of your comments. Please re-review?

@dotnet/roslyn-compiler May I please have a second review?

File: src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs [hunk 4]
```diff
@@ -555,15 +555,21 @@ private SymbolInfo TypeFromLocal(VariableDeclarationSyntax variableDeclaration,
...
+        /// Figures out if this expression is a type in an out-var ArgumentSyntax.
         /// Outputs the VariableDeclarationSyntax directly containing it, if that is the case.
         /// </summary>
         private static bool IsOutVarType(ExpressionSyntax expression, out SyntaxNode parent)
         {
-            return (parent = expression.Parent)?.Kind() == SyntaxKind.VariableDeclaration &&
-                     ((VariableDeclarationSyntax)parent).Type == expression &&
-                     (parent = parent.Parent)?.Kind() == SyntaxKind.Argument &&
-                     ((ArgumentSyntax)parent).Type == expression;
+            parent = null;
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs]: Assignment can be moved inside `if`.

**Comment by jcouv** [comment]
LGTM, thanks

File: src/Features/CSharp/Portable/InternalUtilities/InternalExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+using System.Threading.Tasks;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    internal static class InternalExtensions
+    {
+        public static ITypeSymbol DetermineParameterType(
+            this ArgumentSyntax argument,
+            SemanticModel semanticModel,
+            CancellationToken cancellationToken)
```
Comment (cston) [inline src/Features/CSharp/Portable/InternalUtilities/InternalExtensions.cs]: Why move this from ArgumentSyntaxExtensions.cs?

**Comment by cston** [comment]
LGTM. Much simpler.

File: src/Features/CSharp/Portable/InternalUtilities/InternalExtensions.cs (hunk 5 above)
Comment (gafter) [inline src/Features/CSharp/Portable/InternalUtilities/InternalExtensions.cs]: It is not a general-purpose helper for public consumption.
